lecture_5/homework/app/models/warrior.rb

```diff
@@ -14,4 +14,41 @@ class Warrior < ApplicationRecord
 
   scope :alive, -> { where('death_date IS NULL') }
   scope :dead, -> { where('death_date IS NOT NULL') }
+
+  after_create :update_siege
+  before_update :update_prev_building
+  # cant put same if: :building_id_changed?  since it will always return false
+  after_update  :update_siege_after
+  # after_commit :update_siege_after
+  after_destroy :update_siege
+
+  def update_siege
+    # call after warrior already was asigned to building list after create,update
+    # works for delete as well
+    Reports::SiegeReport.call(building: building) if building
+  end
+
+  # it seems like rspec didnt invoke those callbacks
+  # when i was testing manually it works
+  # probably messed up in factory and tests :(
+  def update_siege_after
+    if prev_building_id
+      # if warrior changed his building run this on both
+      # since in that moment he is no longer in old list , rather in new building
+      # so this way i can keep both old and new buildings updated
+      # puts "prev building: #{prev_building_id} current: #{building_id}"
+      Reports::SiegeReport.call(building: Building.find(prev_building_id))
+      Reports::SiegeReport.call(building: Building.find(building_id))
+      # set this to nil to prevent callback invoking from random update
+      update_column(:prev_building_id, nil)
+    end
+  end
+
+  def update_prev_building
+    # if warrior is changing his buildings
+    if building_id_was != building_id
+      # save this attribute wihtout triggering callbacks
+      update_column(:prev_building_id, building_id_was)
```

Comment:
Z tego co pamiętam update_column omija walidację więc nie jest to najlepszy sposób na edycje kolumny :D Najprosciej byłoby dodać relację typu belongs_to z opcją optional: true :) Optional pozwoli na ustawianie nila na tej kolumnie, nie trzeba nic zmieniać w bazie, bo bazy domyślnie pozwalają na nile, a jedynie rails sam waliduje to że nie mogą być nilem :)

Reply:
Poza tym w sumie gdy sie wbijasz callbackiem miedzy zmiane danych a sve, masz dostep do zmiennej metody changes, ktora zwraca to co sie zmienilo, dzieki czemu nie ma potrzeby w tym przypadku poprzednim budynek na uzytkowniku, bo w callbacku informacja z jakiego na jakiego zmienilo sie building id juz jest